Source/Core/Core/MemTools.cpp

```diff
@@ -264,10 +267,38 @@ static void sigsegv_handler(int sig, siginfo_t* info, void* raw_context)
                                  ))
   {
     // retry and crash
-    signal(SIGSEGV, SIG_DFL);
-#ifdef __APPLE__
-    signal(SIGBUS, SIG_DFL);
-#endif
+    // According to the sigaction man page, if sa_flags "SA_SIGINFO" is set to the sigaction
+    // function pointer, otherwise sa_handler contains one of:
+    // SIG_DEF: The 'default' action is performed
+    // SIG_IGN: The signal is ignored
+    // Any other value is a function pointer to a signal handler
+
+    struct sigaction* old_sa;
+    if (sig == SIGSEGV)
+    {
+      old_sa = &old_sa_segv;
+    }
+    else
+    {
+      old_sa = &old_sa_bus;
+    }
+
+    if (old_sa->sa_flags & SA_SIGINFO)
```

Comment:
What if there were no prior signal handlers registered, would the `oldsig` contain a sane value in those cases?
I never worked with that before, but it looks slightly dodgy to me...

Reply:
I believe it's defined, as the way it signals there was no previous registered handler is .sa_handler == SIG_DFL?
At least in Bionic's source code the old_sa is always set to the kernel-returned values, and they are always initialised.